### PR TITLE
Agile cost forecast: Kraken billing period, AgilePredict forecast, cost_forecast job

### LINKED ISSUE #420: Kraken GraphQL client — billing period lookup
## Parent

Spec: `.agent-docs/specs/feature-agile-cost-forecast.md`

## What to build

A new small GraphQL client, sibling to `OctopusTransport` (`app/data/octopus/transport.py`) rather than an extension of it — different host (`https://api.octopus.energy/v1/graphql/`), different wire format (JSON-over-POST, not query-string GET), no response Pydantic model reuse from the REST clients. Mint a short-lived JWT per call via `obtainKrakenToken(input: {APIKey: "<existing_key>"})`, reusing `OctopusAPISettings.api_key` — no new stored secret, no persisted/rotated refresh token (re-authenticate every call). Query `account(accountNumber: "...") { billingOptions { currentBillingPeriodStartDate currentBillingPeriodEndDate isFixed } }`.

Billing period end date logic:

- `isFixed: true` → use `currentBillingPeriodEndDate` directly.
- `isFixed: false` (flexible billing, confirmed live on this account) → `billing_period_end = billing_period_start + 1 calendar month`, same day-of-month, clamped to the last valid day of that month if it doesn't exist (e.g. the 31st rolling back to the 28th/29th/30th). Recomputed fresh from whatever `currentBillingPeriodStartDate` currently is on every call — never cached across runs.

On an unreachable/error Kraken response, raise a clear exception rather than swallowing it — the daily job (a later issue) is responsible for catching this and recording a failed `job_run`; this client's job is only to fail loudly.

## Acceptance criteria

- [ ] New client authenticates via `obtainKrakenToken(APIKey: ...)` and queries `billingOptions` for a given account number
- [ ] `isFixed: true` returns `currentBillingPeriodEndDate` directly as the billing period end
- [ ] `isFixed: false` computes `billing_period_end` as start + 1 calendar month, same day-of-month, clamped to the month's last valid day (tested with a start date on the 29th/30th/31st rolling into a shorter month)
- [ ] An unreachable/error Kraken response raises a clear exception (not silently swallowed or returning a sentinel)
- [ ] No JWT or refresh token is persisted across calls — every call re-authenticates via the stored API key
- [ ] Unit tests (`responses`-mocked GraphQL POST): `isFixed: true` case, `isFixed: false` fallback case (including the month-clamping edge case), and an unreachable/error case
- [ ] Existing test suite remains green

## Blocked by

None

### PR DESCRIPTION
**Pull request by mholubinka1**
## Summary

- **#420** — Kraken GraphQL client for billing period lookup: mints a short-lived JWT per call via the existing API key, fetches `billingOptions`, and applies the flexible-billing (`isFixed: false`) fallback (start + 1 calendar month, clamped to month-end).
- **#421** — `AgilePredictClient` for the 14-day Agile price forecast, a new `agile_forecast` table (backs the pre-existing "Price Curve" Grafana panel), a forecast-tiling pure function for projecting beyond the real forecast horizon, and a consumption-projection pure function.
- **#422** — `cost_forecast` table, `CostForecastRetriever`, and a new daily `cost_forecast_refresh` job tying it all together to compute `actual_cost_to_date` and `projected_total_cost` — makes the pre-existing `cost_forecast`-reading Grafana panels return real data.

Backend-only — the Grafana panels this feature backs already exist from an earlier reconciliation pass; no dashboard changes in this PR.

Closes #420, #421, #422

## Notable review history

This branch went through 8 rounds of two-axis (Standards + Spec) review. Six rounds found and fixed real issues, the most significant being two money-calculation bugs in the remaining-cost projection:
- The Agile branch's remaining-day window used an exclusive upper bound at the billing period's end date, silently excluding the entire final billable day's forecast slots from the cost projection.
- The fixed/variable-tariff branch used a whole-day count that excluded "today" entirely, so today's not-yet-metered remaining hours never got priced — a systematic undercount on every daily run, since Octopus consumption data lags and "today" often has zero rows by the time the 04:00 job runs.

Both were verified by direct execution (re-deriving the date/hour arithmetic with `uv run python`) and by temporarily reverting each fix to confirm the corresponding regression test actually catches it, not just passes coincidentally.

## Test plan

- [x] 154 tests passing, full suite (up from 103 at branch start)
- [x] Pre-commit clean on both changed-files and full-repo passes (mypy, black, isort, ruff, pylint, bandit, secrets-detection)
- [x] All 10 Given-When-Then scenarios agreed via a Three Amigos session for #422 have direct test coverage
- [x] AgilePredict client schema verified against the live API (not just documentation) before implementation
- [x] Two-axis code review to two consecutive clean passes on both axes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/data/mysql/client.py [hunk 1]
```diff
@@ -210,6 +217,126 @@ def write_product_rate(
...
+            row = (
+                session.query(pr)
+                .filter(
+                    pr.product_code == product_code,
+                    pr.region == region,
+                    pr.valid_from <= as_of,
+                    or_(pr.valid_to.is_(None), as_of < pr.valid_to),
+                )
+                .first()
+            )
```
[the remark this answers is not in the document]
Comment (mholubinka1) [reply app/data/mysql/client.py:269]: Fixed. Added explicit `.order_by(pr.valid_from.desc())` before `.first()`, so an overlapping-rows scenario deterministically prefers the most-recently-started rate (same convention already used for current-rate lookups in grafana/mariadb/queries.md). Added a regression test seeding two overlapping open-ended rows to confirm.

File: app/data/mysql/client.py [hunk 2]
```diff
@@ -210,6 +217,132 @@ def write_product_rate(
...
+                        or_(a.valid_to.is_(None), c.period_from < a.valid_to),
+                    ),
+                )
+                .join(
+                    pr,
+                    and_(
+                        pr.product_code == a.product_code,
+                        c.period_from >= pr.valid_from,
+                        or_(pr.valid_to.is_(None), c.period_from < pr.valid_to),
+                    ),
```
[the remark this answers is not in the document]
Comment (mholubinka1) [reply app/data/mysql/client.py:316]: Fixed. Added a `region: str` parameter to `read_elapsed_billing_period_costs`, threaded through the CostForecastSource Protocol/MonitoringClient/test adapter, and added `pr.region == region` to the product_rate join condition. Confirmed the bug was real and severe by temporarily reverting: kWh doubled (2.0 -> 4.0) and cost inflated ~4x (£0.88 -> £3.37) with a second region's overlapping rate present. New regression test: test_a_rate_for_another_region_is_not_double_matched.

File: app/data/cost_forecast.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                midday = _midnight_utc(day) + timedelta(hours=12)
+                rate = self._client.read_current_product_rate(
+                    agreement.product_code, self._client.region_code, midday
+                )
+                if rate is None:
+                    logger.warning(
+                        f"No product_rate found for {agreement.product_code} "
+                        f"on {day} -- omitting its standing charge from "
+                        "actual_cost_to_date."
+                    )
```
[the remark this answers is not in the document]
Comment (mholubinka1) [reply app/data/cost_forecast.py:198]: Fixed. Changed the log-and-continue to raise RuntimeError, matching this file's established philosophy of failing loudly rather than silently producing a wrong money figure (the current-rate lookup in _project_remaining_cost already does this). New test: test_no_product_rate_for_a_zero_consumption_elapsed_day_raises_and_writes_no_row, using two carefully-positioned rate rows to isolate this code path from the separate remaining-cost rate lookup.
